Big clean up +  frame standarization

husarafal opened the pull request:
These frame id changes are pretty invasive. Please test the system thouroughly, and verify that sensors work correctly and data is in proper format.

In order to make the code review process easier, I divided my work into multiple commits, each related to a different bulk change.

---

## Changelog:

* Fixed Teltonika URDF

  * It was shutting down without `device_namespace`
  * Turned out there was a frame named `device_namespace`
  * Standardized with Ouster and now it works

* Added disclaimer about manipulators

* Added DEVs to README

* Changed `namespace` arguments in URDFs to `robot_namespace` to avoid confusion
  (btw, this is an extra launch argument, not present in `components.yaml`)

* Removed occurrences of `component_name` in unit tests to prepare for renaming `device_namespace`

* In `components.urdf.xacro`, if `device_ns` (now being changed to `component_name`) was empty, for some sensors it was replaced with defaults

  * Removed those defaults for oak, zed, CAM02 `"camera"`, CAM01 `"camera"`

* Replaced `device_namespace` with `component_name`, and updated corresponding YAML queries to `component["name"]`

* Changed `device_ns` in `components.urdf.xacro` to `component_name`, since it’s exactly the same value and there are no conflicts requiring an abbreviated form

* Updated macro param list in `README.md`

* Removed default values of `component_name` in macros

* In macros renamed the helper properties to:
  * `ns` → `robot_namespace_slash`
  * `prefix` → `component_name_underscore`
  * `device_ns` → `component_name_slash`

* Replaced usage of `${robot_namespace}/` and `${component_name}_` or with `/` at the end

* Added missing robot prefix to `gazebo` `gz_frame_id` in Kinova macro

* Added default component names in macros

* Frame unification:

  * Ensured that Gazebo reference frames don’t have the `robot_namespace` prefix
  * Where there was a custom root frame, set the name to `${component_name}_link` – decided that the base frame of a sensor should be named `${component_name}_link`
  * Root frame = mount point is now `{component_name}_link`. If this point differs from the sensor’s origin, then we add a new `{component_name}_origin_frame` / `_laser_frame` / `_camera_center_frame` / ...
  * For example, renamed `orbbec_astra_link` → `astra_visual_frame`

husarafal committed:
readme: add disclaimer that manipulators are only supported in sim

husarafal committed:
readme: add DEVs to sensor table

husarafal committed:
code style: align sensor table rows

husarafal committed:
fix teltonika URDF crash

husarafal committed:
rename unused "name" param in kinova to avoid confusion with component names

husarafal committed:
urdf: rename namespace arg to robot_namespace

husarafal committed:
remove all existing occurences of "component_name"; component_name -> component_model_name

husarafal committed:
components.urdf: remove default component names

husarafal committed:
device_namespace -> component_name

husarafal committed:
readme: update xacro params list

husarafal committed:
component-specific urdfs: remove default component names

husarafal committed:
main urdf: ns -> robot_namespace_value

husarafal committed:
macros: rename prefixes:
ns -> robot_namespace_slash
prefix -> component_name_underscore
device_ns -> component_name_slash

husarafal committed:
macros: replace the uses of ${robot_namespace}/ and ${component_name}_ or with / at the end

husarafal committed:
kinova: add robot_namespace_slash gz_frame_id prefix

husarafal committed:
macros: add default component names

husarafal committed:
REVIEWME: standarize frame names

husarafal commented:
Tests are failing because the new frames are not consistent with the previous standard.

The biggest problem is that I didn't see the need to include manufacturer details in gazebo sensor names. I'm not sure if I was supposed to do that, whether I should revert or just modify the test file.

Another reason is that many base frames have changed from formats like `laser` to `laser_frame`, etc. I think modifying test file here is the way to go.

Final cause of failure: I have introduced default names for all components, so that if `slamtec_rplidar_s1` is instantiated without `component_name`, it will be named `rplidar` by default. But without component name, the test expects `laser_frame` to exist, and not `rplidar/laser_frame`. I'm convinced to keep my default names solution as we agreed to rename `device_namespace` -> `component_name`, and while namespaces can naturally be empty, names generally should not.

husarafal commented:
Hey @KmakD @rafal-gorecki,
Please provide an opinion on my refactoring choices, and help me choose the best way to deal with these test failures.

KmakD commented on launch/gz_components.launch.py:
```diff
@@ -37,17 +37,17 @@ def get_value(node: yaml.Node, key: str):
 
     gz_bridge_name_prefix = component["type"] + "_gz_bridge"
-    device_namespace_prefix = get_value(component, "device_namespace")
+    device_namespace_prefix = get_value(component, "name")
```
This has the same value as `component_name` defined in line 40, can be removed

husarafal replied:
This is a tricky one. Notice that component_name includes a leading slash in case of launch files other than for ur, kinova and robotiq. It is inconsistent in a sense that different per-component gz bridge launch files expect robot_namespace and component_name with or without leading slash instead of all having a homogeneous interface.

Additionally the gz bridge launch files often provide defaults for the component name if it is not provided in `components.yaml`, which are separate and different from defaults in URDF. :/ This clearly needs an overhaul, but that would make the refactor even more complex and would require more testing.

What I can do for now is just rename `device_namespace_prefix` -> `component_name_gz_prefix`

husarafal replied:
@KmakD is the simple rename enough or should I perform apply a large restructurization, since we will have to conduct testing anyway at the conclusion of this PR?

I would provide default component names in every component-specific gz bridge launch file and make sure they are the same as in the URDFs. Same thing in two places, but saves on complexity.

I would also make all gz launch files accept robot_namespace and component_name that are supposed to be either "" or "panther", "front_camera" - without leading slash. Make kinova/ur/robotiq be consistent with others. Handle slash/underscore appending in the individual gz bridge launch files.

KmakD replied:
The rename is enough, I missed that component_name is modified afterwards :sweat_smile:

KmakD commented on launch/gz_ur.launch.py:
```diff
@@ -28,39 +28,39 @@
+    # After this fix the component_name and --namespace should be used.
     robot_namespace_ext = PythonExpression(
         ["''", " if '", robot_namespace, "' == '' ", "else ", "'", robot_namespace, "_'"]
     )
```
Looks like the issue was fixed

husarafal replied:
I removed `robot_namespace_ext`, all instances of it, and uncommented `--namespace` in `controller_manager/spawner` nodes

KmakD commented on launch/gz_intel_realsense_d435.launch.py:
```diff
@@ -32,19 +32,19 @@ def generate_launch_description():
         },
     )
 
     declare_device_namespace = DeclareLaunchArgument(
```
```suggestion
    declare_component_name = DeclareLaunchArgument(
```
Applies to all launch files

husarafal replied:
Replaced

KmakD commented on urdf/components.urdf.xacro:
```diff
@@ -41,7 +41,7 @@ limitations under the License.
 
       <xacro:property name="type" value="${component['type']}" scope="parent" />
-      <xacro:property name="ns" value="${robot_namespace}" scope="parent" />
+      <xacro:property name="robot_namespace_value" value="${robot_namespace}" scope="parent" />
```
This property might be removed, as it only copies value of `robot_namespace`

husarafal replied:
Removed

KmakD commented on urdf/kinova.urdf.xacro:
```diff
@@ -57,19 +57,24 @@ limitations under the License.
-
     <!-- robot name parameter -->
-    <xacro:property name="name" value="kinova_arm" />
+    <xacro:property name="robot_name" value="kinova_arm" />
```
Is this property used?

husarafal replied:
Removed

KmakD commented on urdf/orbbec_astra.urdf.xacro:
```diff
@@ -93,57 +99,57 @@ limitations under the License.
-      frame_id="${ns}${device_namespace}_color_optical_frame"
+      reference="${component_name_underscore}link"
+      name="${robot_namespace_slash}${component_name_underscore}orbbec_astra_color"
+      topic="${robot_namespace_slash}${component_name_slash}color/image_raw"
```
We aim to have the default configuration the same as for the real sensor when possible. Astra afaik default topic is `/camera/color/image_raw` so the default `component_name` should also be `camera`. This also applies to other sensors.

husarafal replied:
I updated default names for realsense and astra to `camera`.

Should I also update default name of ZEDs and OAKs to `camera`?

On `ros2` branch:
https://github.com/husarion/husarion_components_description/blob/1a77f7b3d9c942b1560dafc823ab0c4a1f125995/urdf/components.urdf.xacro#L141
https://github.com/husarion/husarion_components_description/blob/1a77f7b3d9c942b1560dafc823ab0c4a1f125995/urdf/components.urdf.xacro#L167

KmakD replied:
I would leave it as is

KmakD commented on urdf/robotiq.urdf.xacro:
```diff
@@ -22,16 +22,20 @@ limitations under the License.
-
-    <xacro:gripper.robotiq_gripper name="${device_namespace}robotiq_${robotiq_type}"
-      prefix="${prefix}"
+    <xacro:gripper.robotiq_gripper name="${component_name_underscore}robotiq_${robotiq_type}"
```
```suggestion
    <xacro:gripper.robotiq_gripper name="${component_name_underscore}${robotiq_type}"
```
otherwise it will result in `robotiq_robotiq_type` name

husarafal replied:
Removed

KmakD commented on urdf/teltonika_003R-00253.urdf.xacro:
```diff
@@ -21,19 +21,25 @@ limitations under the License.
 
-    <xacro:property name="ns" value="${namespace + '/' if namespace != '' else ''}" />
+    <xacro:if value="${component_name == ''}">
+      <xacro:property name="component_name" value="teltonika" />
```
```suggestion
      <xacro:property name="component_name" value="gps" />
```
to have the sim default topic match the real sensor

husarafal replied:
Updated

KmakD reviewed (changes requested):
Overall great job. There are minor fixes. 

As for tests, ideally we would like to keep the topic and frame_id names the same as the real hardware driver. If this is impossible, because it would require to heavily break the standard, only then I would modify the test files

husarafal committed:
device_namespace_prefix -> component_name_gz_prefix

husarafal committed:
remove ros2_control bug workaround
https://github.com/ros-controls/ros2_control/issues/1506

husarafal committed:
replace leftover device_namespace -> component_name

husarafal committed:
main urdf: remove robot_namespace_value

husarafal committed:
kinova urdf: remove robot_name

husarafal committed:
urdfs: set default d435 and astra names to "camera"

husarafal committed:
Remove extra robotiq_ in name

husarafal committed:
urdf default name teltonika -> gps

husarafal committed:
revert ouster/velodyne/rplidar/zed reading frame changes

husarafal commented:
To address the topic/frame hardware consistency I have reverted my changes:
- ouster: `{name}_lidar_frame` -> `{name}_os_lidar`
- rplidar: `{name}_laser_frame` -> `{name}_laser`
- zed: `{name}_camera_center_frame` -> `{name}_camera_center`
- velodyne: `{name}_lidar_frame` -> `{name}_velodyne`

So only the problematic sensors that did not pass the tests.

I did not revert:
- teltonika `{name}_origin_frame` -> `{name}` (Just `{name}` was a very weird choice for the frame.)

husarafal commented:
I'm still inclined to keep my changes to gazebo sensor names and modify the test file instead. We don't necessarily need to include full manufacturer and model information in the sensor names, and some of our URDFs don't even have our custom gz_sensor definitions, so we can't control the gazebo sensor names of those. No need to worry about unique naming too, because `component_name` prefix is included in the gazebo sensor name. I would modify the test file to remove gazebo sensor name checks

husarafal commented:
Regarding the failing component_name="" urdf instantiation tests, I remain convinced that default component names should be in place for all types of sensors. I'm inclined towards modifying the test file to prevent it from checking `does_sensor_name_exist(<frame name but without component_name prefix>)` in the component_name="" pass

husarafal commented:
@KmakD Further comment required

KmakD commented:
Regarding teletonika, If we change the dafualt (which in deed seems weird), we would have to handle that in teletonika launch accordingly so the real sensor still works. Probably [here](https://github.com/husarion/husarion_ugv_ros/blob/702a1bd20f6c13ecff038e9df55e92a0c13b45f9/husarion_ugv_localization/launch/nmea_navsat.launch.py#L62)

I'm ok, with keeping sensor names as you suggested, just modify the test files where needed.

The default components names should be defined for all sensors (dev doesn't need that), tests can be updated to mach new standard.

husarafal committed:
standarize astra and zed gz sensor names

husarafal committed:
fix unit tests

husarafal committed:
remove unused imports

husarafal committed:
rename teltonika frame

husarafal commented:
I have updated the teltonika frame in `husarion_ugv_ros`. Also removed all occurences of `device_namespace`.

https://github.com/husarion/husarion_ugv_ros/pull/611

husarafal commented:
@KmakD Thanks for your guidance. All our points have been addressed, tests are passing, and the PR is ready for possible additional review and testing.

husarafal committed:
Merge branch 'ros2' into multi-fix
